join.go

```diff
@@ -0,0 +1,26 @@
+package gomatrixserverlib
+
+import (
+	"context"
+
+	"github.com/matrix-org/gomatrixserverlib/spec"
+)
+
+type FederatedJoinClient interface {
+	MakeJoin(ctx context.Context, origin, s spec.ServerName, roomID, userID string) (res MakeJoinResponse, err error)
+	SendJoin(ctx context.Context, origin, s spec.ServerName, event *Event) (res SendJoinResponse, err error)
```

Comment:
This is okay for now, but I'm wondering if we can be less specific long term. If we ever decide to change the flow here (removing make_join, adding a third stage) then everything needs to change which is unfortunate. It would be nice to hide this behind the room version PDU bits longer term I guess?